### DISCUSSION TITLE
Guardian: Create message-broker plugin capability & write Google PubSub plugin

### DISCUSSION THREAD

**Discussion by voycey** [hiero-ledger/hiero-improvement-proposals#287] [Ideas]
There are a number of items that need to be addressed in the Guardian before it can be made production ready, one of the main items is that FastMQ which is currently in use has no TLS component to it or authentication method.

In situations where data needs to be sent from outside the VPC this represents a security deficiency that needs to be addressed.

For our purposes we plan to implement integration with Google PubSub and the Google PubSub emulator (for local use), the implementation would look like:

1. Create an interface for the message broker that includes:
    - /mrv endpoint (publish)

2. Modify FastMQ to extend this interface and create package `message-broker-fastmq` and modify `package.json` to keep default setup
3. Create new package `message-broker-pubsub`
4. Look at possibly implementing this via webpack (https://webpack.js.org/plugins/copy-webpack-plugin/) to allow for easy customisation
5. PubSubEmulator Docker Container would be created for local development (named `message-broker`)
6. Environment variables for production setup to connect to Live PubSub vs PubSubEmulator

This setup would keep the current default setup but provide a plugin architecture to allow for new message queue plugins to be built according to requirements (AWS SQS & Redis for example).

**Comment by anvabr** [comment]
@voycey, thank you for the review and suggestion above. The team behind Guardian will take it onboard. However this is not the right place to submit bug reports and enhancement/feature requests for Guardian. The best way is to create an issue in the guardian repo https://github.com/hashgraph/guardian. Just click on 'Issues' and 'New' and write-up your notes and Guardian product team will process it accordingly.

**Reply by anvabr** [reply]
Recorded in https://github.com/hashgraph/guardian/issues/215, https://github.com/hashgraph/guardian/issues/216

**Comment by prernaadev01** [comment]
@voycey, Can you please let us know when do you need this setup to be completed, in order to set our dev priorities. Thank you.

**Comment by voycey** [comment]
We are happy to take on this work if there are other priorities you need to focus on? As always we need it ASAP, we can't push into production without it so it's next on our list

**Reply by anvabr** [reply]
Hi @voycey , we would absolutely very much appreciate your team's contribution to the codebase. If you are able to work on this immediately, please go ahead. Our backlog is full and it'd be difficult for us to take this on right now. We'll review your PR when you are ready, and please let us know how else we could help

**Reply by voycey** [reply]
We need to catch up with the latest set of updates first but after that we will work on this feature.
I would also like to discuss collaboration methods for if we want to build additional functionality on the guardian, without a stable API contract it is difficult to be able to make updates and add required functionality we require, is there anything in the pipeline to support this?

**Reply by anvabr** [reply]
> without a stable API contract it is difficult to be able to make updates and add required functionality we require, is there anything in the pipeline to support this?

We are planning to stabilise the API from the upcoming release onwards. Please see [API versioning and deprecation policy.](https://docs.hedera.com/guardian/getting-started/api-versioning-and-depreciation-policy)

**Reply by voycey** [reply]
Ok great - we will submit PRs for changes we think benefit the solution as a whole once we have caught up with the latest codebase
